sympy/solvers/ode.py

```diff
@@ -6625,55 +6625,64 @@ def _linear_2eq_order1_type1(x, y, t, r, eq):
-                raise NotImplementedError('b == 0 case not implemented')
-            gsol1 = r['b']*exp(sigma*t)*(C1*sin(beta*t)+C2*cos(beta*t))
-            gsol2 = exp(sigma*t)*(((C1*(sigma-r['a'])-C2*beta)*sin(beta*t)+(C1*beta+(sigma-r['a'])*C2)*cos(beta*t)))
-        if D == 0:
-            if r['a']!=r['d']:
-                gsol1 = 2*r['b']*(C1 + C2/(r['a']-r['d'])+C2*t)*exp((r['a']+r['d'])*t/2)
-                gsol2 = ((r['d']-r['a'])*C1+C2+(r['d']-r['a'])*C2*t)*exp((r['a']+r['d'])*t/2)
-            if r['a']==r['d'] and r['a']!=0 and r['b']==0:
-                gsol1 = C1*exp(r['a']*t)
-                gsol2 = (r['c']*C1*t+C2)*exp(r['a']*t)
-            if r['a']==r['d'] and r['a']!=0 and r['c']==0:
-                gsol1 = (r['b']*C1*t+C2)*exp(r['a']*t)
-                gsol2 = C1*exp(r['a']*t)
-    elif (r['a']*r['d'] - r['b']*r['c']) == 0 and (r['a']**2+r['b']**2) > 0:
-        k = r['c']/r['a']
-        if r['a']+r['b']*k != 0:
-            gsol1 = r['b']*C1 + C2*exp((r['a']+r['b']*k)*t)
-            gsol2 = -r['a']*C1 + k*C2*exp((r['a']+r['b']*k)*t)
+    C1, C2 = get_numbered_constants(eq, num=2)
+    a, b, c, d = r['a'], r['b'], r['c'], r['d']
+    real_coeff = all(v.is_real for v in (a, b, c, d))
+    D = (a - d)**2 + 4*b*c
+    l1 = (a + d + sqrt(D))/2
+    l2 = (a + d - sqrt(D))/2
+    equal_roots = Eq(D, 0).simplify()
+    gsol1, gsol2 = [], []
+
+    # Solutions have exponential form if either D > 0 with real coefficients
+    # or D != 0 with complex coefficients. Eigenvalues are distinct.
+    # For each eigenvalue, pick an eigenvector, making sure we don't get (0, 0)
+    assumption = D > 0 if real_coeff else Not(equal_roots)
+    vector1 = Matrix((l1 - d, c)) if b == l1 - a == 0 else Matrix((b, l1 - a))
+    vector2 = Matrix((l2 - d, c)) if b == l2 - a == 0 else Matrix((b, l2 - a))
+    sol_vector = C1*exp(l1*t)*vector1 + C2*exp(l2*t)*vector2
+    gsol1.append((sol_vector[0], assumption))
+    gsol2.append((sol_vector[1], assumption))
+
+    # Solutions have trigonometric form for real coefficients with D < 0
+    # Both b and c are nonzero in this case, so (b, lam-a) is an eigenvector
+    # It splits into real/imag parts as (b, sigma-a) and (0, beta). Then
+    # multiply it by C1(cos(beta*t) + I*C2*sin(beta*t)) and separate real/imag
+    assumption = D < 0 if real_coeff else False
+    sigma = re(l1)
+    if im(l1).is_positive:
+        beta = im(l1)
+    else:
+        beta = im(l2)
+    vector1 = Matrix((b, sigma - a))
+    vector2 = Matrix((0, beta))
+    sol_vector = exp(sigma*t) * (C1*(cos(beta*t)*vector1 - sin(beta*t)*vector2) + \
+        C2*(sin(beta*t)*vector1 + cos(beta*t)*vector2))
+    gsol1.append((sol_vector[0], assumption))
+    gsol2.append((sol_vector[1], assumption))
+
+    # Final case is D == 0, a single eigenvalue. If the eigenspace is 2-dimensional
+    # then we have a scalar matrix, deal with this case first.
+    assumption = equal_roots
+    if a == d and b == c == 0:
```

Comment:
You have to be careful when making decisions based on equalities since mathematically identical variants of a quantity will fail such tests. You've made good use of Piecewise...might you make `sol_vector` itself a Piecewise:

```
cond1 = And(Eq(a, d), Eq(b, c), Eq(c, 0))
cond2 = And(Eq(b, 0), Eq(l1, a))
M = lambda a, b: Matrix((a, b))
vector1 = Piecewise(
    (M(1, 0), cond1), 
    (M(l1 -d, c), cond2), 
    (M(b, l1 - a), True))
vector2 = Piecewise(
    (M(0, 1), cond1),
    (M(1, 0), cond2), (Piecewise(
        (M(0, 1), Eq(a, l1)),
        (M(b/(a - l1), 0), True)), True))
```

Reply:
> .might you make sol_vector itself a Piecewise:

I mean, "it might be good if you make `sol_vector` itself a Piecewise as shown:"

Reply:
I will, but following further on this, I think I should work more on this and make it so that the solution for each of two functions is a single flat Piecewise, avoiding sums of Piecewise objects and nested Piecewise.

Reply:
`piecewise_fold` will flatten the expression.

Reply:
A more serious obstruction is that sol_vector[0] will not work. TypeError: 'Piecewise' object does not support indexing

Reply:
A possible solution to the latter is to devise some way of folding Piecewise expressions with Matrix objects into a Matrix object with Piecewise expressions. 

Another, which I think is easier, is to replace the checks like a==0 by `a.equals(0)`. It's expensive, but (1) only for complicated symbolic systems; (2) the wait will be worth it if the answer isn't a long combination with nested Piecewise formulas.

Reply:
> I mean, "it might be good if you make sol_vector itself a Piecewise as shown:"

I mean, "make the individual vector1 and vector2 to be flat Piecewise as shown." That would take care of the indexing issue. The "flat" result is not shown in the snippet. You would have to do `sol_vector = [piecewise_fold(i) for i in (vector1, vector2)]`.

Regarding using `equals`...this is where my preference is to quickly return what you can so the user can decide if they want to try to simplify things further, e.g. by using `equals` on an Eq's args.

Reply:
On the other hand, the output of this solver does not necessarily go directly to the user. If there are initial conditions, it will go to `solve` and hit #13203 there, as recently reported in #14001.

Reply:
Let's just do this part right. If the form of the solution depends on these conditions, then we don't want to introduce a situation where the wrong solution is selected because a relationship could not be determined. If the situation is easily resolved, the Piecewise will vanish; and if not, we at least get the right solution, waiting for enough information to select the correct branch.